backend/module/chaos/serverexperimentation/xds/ecds.go

```diff
@@ -0,0 +1,302 @@
+	gcpTypes "github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	proto "github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/duration"
+	"go.uber.org/zap"
+
+	experimentation "github.com/lyft/clutch/backend/api/chaos/experimentation/v1"
+	serverexperimentation "github.com/lyft/clutch/backend/api/chaos/serverexperimentation/v1"
+)
+
+const (
+	envoyDownstreamServiceClusterHeader   = `x-envoy-downstream-service-cluster`
+	faultFilterConfigNameForExternalFault = `envoy.egress.extension_config.%s`
+	faultFilterConfigNameForInternalFault = `envoy.extension_config`
+	faultFilterTypeURL                    = `type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault`
+
+	delayPercentRuntime    = `ecds_runtime_override_do_not_use.http.delay.percentage`
+	delayDurationRuntime   = `ecds_runtime_override_do_not_use.http.delay.fixed_duration_ms`
+	abortHttpStatusRuntime = `ecds_runtime_override_do_not_use.http.abort.http_status`
+	abortPercentRuntime    = `ecds_runtime_override_do_not_use.http.abort.abort_percent`
+)
+
+// Serialized default fault filter
+var SerializedDefaultFaultFilter []byte
+
+// Default abort and delay fault configs
+var DefaultAbortFaultConfig = &gcpFilterFault.FaultAbort{
+	ErrorType: &gcpFilterFault.FaultAbort_HttpStatus{
+		HttpStatus: 503,
+	},
+	Percentage: &gcpType.FractionalPercent{
+		Numerator:   0,
+		Denominator: gcpType.FractionalPercent_HUNDRED,
+	},
+}
+
+var DefaultDelayFaultConfig = &gcpFilterCommon.FaultDelay{
+	FaultDelaySecifier: &gcpFilterCommon.FaultDelay_FixedDelay{
+		FixedDelay: &duration.Duration{
+			Nanos: 1000000, // 0.001 second
+		},
+	},
+	Percentage: &gcpType.FractionalPercent{
+		Numerator:   0,
+		Denominator: gcpType.FractionalPercent_HUNDRED,
+	},
+}
+
+type SafeEcdsResourceMap struct {
+	mu                    sync.Mutex
+	requestedResourcesMap map[string][]string
+}
+
+func (safeResource *SafeEcdsResourceMap) getResourcesFromCluster(cluster string) []string {
+	var resourceNames []string
+
+	safeResource.mu.Lock()
+	if _, exists := safeResource.requestedResourcesMap[cluster]; exists {
+		resourceNames = safeResource.requestedResourcesMap[cluster]
+	}
+	safeResource.mu.Unlock()
+
+	return resourceNames
+}
+
+func (safeResource *SafeEcdsResourceMap) setResourcesForCluster(cluster string, newResources []string) {
+	existingResources := safeResource.getResourcesFromCluster(cluster)
+
+	safeResource.mu.Lock()
+	defer safeResource.mu.Unlock()
+
+	var missingResources []string
+	for _, newResource := range newResources {
+		isMissing := true
+		for _, resource := range existingResources {
+			if newResource == resource {
+				isMissing = false
+				break
+			}
+		}
+
+		if isMissing {
+			missingResources = append(missingResources, newResource)
+		}
+	}
+
+	safeResource.requestedResourcesMap[cluster] = append(safeResource.requestedResourcesMap[cluster], missingResources...)
```

Comment:
Should we just use a map[string]struct[] here? That way we could just unconditionally add all the resources